Donation-platform/src/main/java/com/google/UpdateTrendingScores.java

```diff
@@ -0,0 +1,113 @@
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.model.Charity;
+import com.google.model.Tag;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.PreparedQuery.TooManyResultsException;
+
+// File that includes DB updates for trends scores
+// Note: Only runs as a result of the offline script
+public final class UpdateTrendingScores extends CharityUtils {
+
+    public UpdateTrendingScores(DatastoreService ds) {
+        super(ds);
+    }
+
+    // returns the trending score of inputted charity calculated
+    // as a weighted average of the tagScore and the avgReview where
+    // tagScore represents the average trending score of the associated tags
+    // and avgReview is a weighted average of the userRating and the charityNavigatory API rating
+    // Note: weights for the averages are stored as class constants
+    private double calcCharityTrendingScore(Charity charity) {
+        boolean hasCharityNavRating = hasCharityNavRating(charity);
+        double charityNavRating = 0;
+        if (hasCharityNavRating) {
+            charityNavRating = calcCharityNavRating(charity);
+        }
+        double userRating = charity.getUserRating();
+        double avgReview;
+        if (hasCharityNavRating) {
+            avgReview = USER_RATING_WEIGHT * userRating + CHARITY_NAV_WEIGHT * charityNavRating;
+        } else {
+            avgReview = userRating;
+        }
+        Collection<Tag> tags = new ArrayList<>();
+        tags = charity.getCategories();
+        double charityTagsScore = 0;
+        try {
+            charityTagsScore = getTagTrendingScore(tags);
+        } catch (EntityNotFoundException e) {
+            System.out.println("tag entity not found: " + e);
+        } catch (TooManyResultsException e) {
+            System.out.println("duplicate tags exist in db: " + e);
+        } catch (Exception e) {
+            System.out.println("unexpected exception: e");
+        }
+        double charityTrendingScore =
+            TAGS_SCORE_WEIGHT * charityTagsScore + AVG_REVIEW_WEIGHT * avgReview;
+        return charityTrendingScore;
+    }
+
+    //update all charity trending scores
+    public void updateCharityScores() {
+        for (Charity charity : charities) {
+            double charityScore = calcCharityTrendingScore(charity);
+            charity.setTrendingScoreCharity(charityScore);
+            try {
+                db.updateCharity(charity);
+            }
+            catch (Exception e) {
+                System.out.println("unable to update charity: " + e);
+            }
+        }
+    }
+
+    // TODO: Decide whether we will be using charityNavRating and change accordingly
```

Comment:
have we still not decided? i thought now we already have these ratings?

Reply:
we have decided that we will load them in, but the changes to add charity navigator rating as an attribute of a Charity and the script changes required to load them into the db are in a separate branch that Ryan is going to be submitting as a PR, so I did not want to address that until those changes are merged, otherwise I'd have to duplicate those changes on my branch to get it to compile.